runtime/guard-jwt/src/main/java/io/aklivity/zilla/runtime/guard/jwt/internal/JwtGuardHandler.java

```diff
@@ -54,39 +66,48 @@ public class JwtGuardHandler implements GuardHandler
     private final Long2ObjectHashMap<JwtSessionStore> sessionStoresByContextId;
 
     public JwtGuardHandler(
-        JwtOptionsConfig options,
-        LongSupplier supplyAuthorizedId)
+            JwtOptionsConfig options,
+            LongSupplier supplyAuthorizedId)
+    {
+        this(options, supplyAuthorizedId, null, null);
+    }
+
+    public JwtGuardHandler(
+            JwtOptionsConfig options,
+            LongSupplier supplyAuthorizedId,
+            Function<URL, String> readURL,
+            Jsonb jsonb)
     {
         this.issuer = options.issuer;
         this.audience = options.audience;
         this.challenge = options.challenge.orElse(null);
 
-        Map<String, JsonWebKey> keys = new HashMap<>();
+        Map<String, JsonWebKey> resolvedKeys = new HashMap<>();
         if (options.keys != null)
         {
-            for (JwtKeyConfig key : options.keys)
+            resolvedKeys = getKeys(options.keys);
+        }
+        else if (options.keysURL.isPresent())
+        {
+            try
             {
-                try
-                {
-                    Map<String, Object> params = new HashMap<>();
-                    params.put("kty", key.kty);
-                    params.put("kid", key.kid);
-                    params.put("e", key.e);
-                    params.put("n", key.n);
-                    params.put("alg", key.alg);
-                    params.put("crv", key.crv);
-                    params.put("x", key.x);
-                    params.put("y", key.y);
-                    params.put("use", key.use);
-                    keys.put(key.kid, JsonWebKey.Factory.newJwk(params));
-                }
-                catch (JoseException ex)
-                {
-                    rethrowUnchecked(ex);
-                }
+                String keysText = readURL.apply(URI.create(options.keysURL.get()).toURL());
+                JsonObject keysObject = jsonb.fromJson(keysText, JsonObject.class);
+                List<JwtKeyConfig> keysConfig = keysObject
+                        .getJsonArray(KEYS_NAME)
+                        .stream()
+                        .map(JsonValue::asJsonObject)
+                        .map(new JwtKeyConfigAdapter()::adaptFromJson)
+                        .collect(toList());
+                resolvedKeys = getKeys(keysConfig);
```

Comment:
We actually want to be able to call something like the following:
```
JwtKeySetConfig jwks = jsonb.fromJson(keysText, JwtKeySetConfig.class);
```
Where `JwtKeySetConfig` represents the syntax of the response from the keysURL, and has a field called `keys` of type `List<JwtKeyConfig>`.
This means we'd need to create a local `jsonb` properly configured with the corresponding (new) class `JwtKeySetConfigAdapter` that encapsulates the parsing logic above, making use of `JwtKeyConfigAdapter` internally to parse each key.

Then we don't need to pass the global `jsonb` here any more either.

Initialization of the local `jsonb` would follow something like:
```
            JsonbConfig config = new JsonbConfig()
                    .withAdapters(new JwtKeySetConfigAdapter());
            Jsonb jsonb = JsonbBuilder.newBuilder()
                    .withConfig(config)
                    .build();
```

Reply:
this is a neat insight!